src/metabase/util/malli/schema.clj

```diff
@@ -333,7 +334,13 @@
      [:slug {:optional true} :string]
      [:name {:optional true} :string]
      [:default {:optional true} :any]
-     [:sectionId {:optional true} NonBlankString]]
+     [:sectionId {:optional true} NonBlankString]
+     [:temporal_units {:optional true}
+      [:sequential
+       (into [:enum]
+             (concat
+              lib.schema.temporal-bucketing/temporal-bucketing-units
+              (map name lib.schema.temporal-bucketing/temporal-bucketing-units)))]]]
```

Comment:
I did it this way because even though the schema `::lib.schema.temporal-bucketing/unit` has a `:decode/normalize` property, it isn't working.

What i mean is, parameters come into this validation with something like this:
`{:temporal_unit ["day" "week"]}`.

While I could transform it during validation using decode and the json-transformer to make it pass here, I get failures in defendpoint. For example: 
https://github.com/metabase/metabase/blob/514564290f02a6d1e4cbdaa0a0b1ffb32ead553c/src/metabase/api/dashboard.clj#L789

It seems that our endpoint macro isn't decoding with the schema, it's just validating against it. This means that I can't use the ::unit schema directly, as it expects keywords and I can only get strings into it.

Reply:
I am concerned about not decoding in `defendpoint`. A primary purpose of putting schemas on `defendpoint` is to normalize things into the canonical form the code expects.

If using `decode` in defendpoint is causing too many failures, it would be okay for this change to wait for that to be fixed up. On the other hand, I'm not very familiar with `defendpoint`'s internals, or our API request pipeline generally, so I might be overlooking important concerns. But in light of the recent quality issues, reducing the number of *Things We Have To Think About* in order to write correct code is always a win, so it would be great to consistently get keywords for units here.

Reply:
Yea, that makes sense to me.
I think the problem is somewhere in defendpoint -> it seems to *not* be running the decode? at least it's not for the :temporal_unit key inside the parameters?

Reply:
Hmmm, I've written `QueryVectorOf` recently and it employs decoding and `defendpoint` handles this just normally. Maybe it's worth taking a look and if you don't know what to do let's team up and try to figure it out? I'd prefer defendpoint schema to be as comprehensive as possible too. 😁

I.e., how about trying out `[:maybe (ms/QueryVectorOf ms/Parameter)]`? 😁

Reply:
I will give a quick blurb about "how to play with defendpoint decoding in the repl" at the next guild sync. It's quite simple really. here is the gist:

```
#_:clj-kondo/ignore ;;nocommit
(require '[malli.core :as mc] '[malli.error :as me] '[malli.util :as mut] '[metabase.util.malli :as mu]
         '[metabase.util.malli.describe :as umd] '[malli.provider :as mp] '[malli.generator :as mg]
         '[malli.transform :as mtx])


;; All coersion on the API layer is controlled by
;; `metabase.api.common.internal/defendpoint-transformer`,
;; which looks like this:
(def tzform (mtx/transformer
             (mtx/string-transformer)
             (mtx/json-transformer)
             (mtx/default-value-transformer)))

(def my-schema [:or {:decode/string (fn [x] (str x "_" x))} :int :string])

(mc/decode my-schema "foo" tzform)
;; => "foo_foo"

(mc/decode my-schema 0 tzform)
;; => "0_0"
```

Reply:
The `:decode/normalize` property will not be used by the `metabase.api.common.internal/defendpoint-transformer` because it doesn't run the transforms named "normalize". So if you want defendpoint coercion of this schema, then there some ways to achieve that:

1) add `(mtx/transformer {:name "noramlize"})` into the defendpoint-transformer stack
2) put a string-transformer onto the `::unit` schema, so it would read like so:
```clojure
(mr/def ::unit
  (into [:enum {:error/message    "Valid temporal bucketing unit"
                :decode/normalize common/normalize-keyword
                :decode/string common/normalize-keyword}]
        temporal-bucketing-units))

```
3) add a new named transformer to defendpoint and use that one so `::unit` would read like so:
```clojure
(mr/def ::unit
  (into [:enum {:error/message    "Valid temporal bucketing unit"
                :decode/normalize common/normalize-keyword
                :decode/defendpoint common/normalize-keyword}]
        temporal-bucketing-units))
```
and the transformer stack:

```clojure
(def metabase.api.common.internal/defendpoint-transformer
  (mtx/transformer 
    (mtx/string-transformer)
    (mtx/json-transformer)
    (mtx/default-value-transformer)
    (mtx/transformer {:name "defendpoint"}))
```